### ISSUE TITLE
Update Gateway to handle auth failures

### ISSUE THREAD

**Issue by mchitre** [org-arl/fjage#167]
Related to #165 

- [x] Java gateway @mchitre 
- [ ] JS gateway @notthetup 
- [ ] Python gateway @prasadtiru 
- [ ] C gateway @prasadtiru 
- [ ] Julia gateway @mchitre

**Comment by mchitre** [comment]
I'll take a stab at the Java one first, and then we can do the rest, once the desired behavior is finalized.

**Comment by mchitre** [comment]
OK, so here's the updated version:
1. Auth failure JSON messages now tell you which message they are replying to.
2. `authenticate()` returns a `boolean` telling us if the auth was successful or not.
3. `SlaveContainer` & `Gateway` now throw `AuthFailureException` if any API that solicits a response (`agentForService`, `agentsForService`, ``) is called, and it fails due to authentication failure.

**Comment by notthetup** [comment]
Yup. That would work for JS. I will update the document with the appropriate information and update the PR.

**Comment by notthetup** [comment]
Doesn't `request` wait for a response from the master container?

**Comment by mchitre** [comment]
No, not at a protocol level.

`request()` is a convenience function that sends a message and listens for incoming messages, filters them, and finds a response. However, at a protocol level, _send_ from slave to master is one primitive, and the _send_ from the master to slave is a different primitive. They don't have the same transaction IDs at the JSON protocol level, but are tied together through `inReplyTo` at a message level.

While one could conceivably look inside the message to figure out the association, it's messy. The auth failure would be associated with the _send_ requests JSON protocol transaction ID, and the response would be associated with the _send_ request's content message's message ID, via the response message's `inReplyTo` field. Handling that in the connection handler, from a design standpoint, merges two different levels of abstractions, and so IMO is undesirable and hacky.

**Comment by notthetup** [comment]
Hmm. The authentications aspect is pretty straightforward.

However, if the message sent from a Gateway/Slave is rejected by the Master container because of an authorization error (`fw.permit` returns `false`), then the Gateway will see it as a timeout and won't be able to differentiate between other errors in the Master container vs authorization error.

We should be able to work around this for the UI, mostly by sinking in the logic of which messages potentially require authorization into the UI and then showing the appropriate errors. Ugly hack but it will work.

But from a general API perspective, not being able to differentiate between an API call being rejected for authorization error vs an API call timing out will potentially make it harder to write applications requiring authentication/authorization.

**Comment by mchitre** [comment]
I agree. The protocol provides the information, just that the Gateway API has trouble exposing it. We could insist that every gateway throw an exception (or equivalent for the language) on a `request()`API as well. It will slightly complicate the implementation of the Gateway, but is doable. So I'm OK to go with that -- I can modify the Java one to do it.

**Comment by mchitre** [comment]
Implemented in 9b4238a0436f469783550ae241b62c2b5c695bb6.

Check only implemented at start and end of `request()`, which means it may wait for a timeout before throwing exception, if the server had not responded by the time the gateway/agent started waiting for response. This is an implementation detail, and other gateways might do it differently, avoiding the wait if they can.

**Comment by notthetup** [comment]
Semi-unrelated to auth failures, but I also noticed the current implementation doesn't block outgoing messages to an unauthenticated Connection. Since internally all messages with `relay` set to true are broadcasted to every Slave container, if a Slave container is connected over a connection and not yet authenticated, it will still receive all broadcasted messages.

**Comment by mchitre** [comment]
Hmm. We should check rules on outgoing too. Will raise an issue.

**Comment by mchitre** [comment]
#182
